oml/retrieval/postprocessors/pairwise.py

```diff
@@ -0,0 +1,206 @@
+import itertools
+from abc import ABC
+from pathlib import Path
+from typing import List, Union
+
+import torch
+from torch import Tensor
+
+from oml.inference.pairwise import (
+    pairwise_inference_on_embeddings,
+    pairwise_inference_on_images,
+)
+from oml.interfaces.models import IPairwiseDistanceModel
+from oml.interfaces.retrieval import IDistancesPostprocessor
+from oml.transforms.images.utils import TTransforms
+from oml.utils.misc_torch import assign_2d
+
+
+class PairwisePostprocessor(IDistancesPostprocessor, ABC):
+    """
+    This postprocessor allows us to re-estimate the *distances* between queries and *top-n* galleries
+    closest to them. It creates pairs of queries and galleries and feeds them to a pairwise model.
+
+    """
+
+    top_n: int
+
+    def process(
+        self, distances: Tensor, queries: Union[List[Path], Tensor], galleries: Union[List[Path], Tensor]
+    ) -> Tensor:
+        """
+        Args:
+            distances: Matrix with the shape of ``[Q, G]``
+            queries: *Q* queries, they may be paths or representations.
+            galleries: *G* galleries, they may be paths or representations.
+
+        Returns:
+            Distance matrix with the shape of ``[Q, G]``,
+            where ``top_n`` minimal values in each row have been updated by the pairwise model,
+            other distances are shifted to keep the relative order.
+
+        """
+        n_queries = len(queries)
+        n_galleries = len(galleries)
+
+        assert list(distances.shape) == [n_queries, n_galleries]
+
+        # 1. Adjust top_n with respect to the actual gallery size and find top-n pairs
+        top_n = min(self.top_n, n_galleries)
+        ii_top = torch.topk(distances, k=top_n, largest=False)[1].view(-1)
+
+        # 2. Create (n_queries * top_n) pairs of each query and related galleries and re-estimate distances for them
+        distances_upd = self.inference(queries=queries, galleries=galleries, ii_top=ii_top, top_n=top_n)
```

Comment:
Here `ii_top` is a flattened array, but in the `self.inference` docstring it is said that `ii_top: Indices of the closest galleries with the shape of [Q, top_n]`. Is it ok to use flattened array in such case?

Reply:
thx for your findings 
code is okay, but I will fix comments

Reply:
thx for your findings
I've aligned shapes and comments for `ii_top` and `distances`